connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/executable/CancellationManager.java

```diff
@@ -0,0 +1,119 @@
+ * distributed with this work for additional information regarding copyright
+ * ownership. Camunda licenses this file to you under the Apache License,
+ * Version 2.0; you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.inbound.executable;
+
+import io.camunda.connector.api.error.RestartException;
+import io.camunda.connector.api.inbound.Health;
+import java.time.Duration;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CancellationManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CancellationManager.class);
+  private final Map<UUID, RegisteredExecutable> executables;
+  private final ScheduledExecutorService reactivationExecutor;
+
+  public CancellationManager(Map<UUID, RegisteredExecutable> executables) {
+    this.executables = executables;
+    this.reactivationExecutor = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  public static CancellationManager create(Map<UUID, RegisteredExecutable> executables) {
+    return new CancellationManager(executables);
+  }
+
+  public Consumer<Throwable> createCancellationCallback(UUID uuid) {
+    return (throwable) -> {
+      switch (throwable) {
+        case RestartException restartException -> {
+          handleCancellation(uuid, throwable);
+          handleRestart(uuid, restartException.getDelay(), restartException.getRetryAttempts());
+        }
+        default -> handleCancellation(uuid, throwable);
+      }
+    };
+  }
+
+  private void handleRestart(UUID uuid, Duration delay, Integer retryAttempts) {
+    LOG.warn(
+        "Inbound connector executable has requested its reactivation in {} {}",
+        delay.getSeconds(),
+        TimeUnit.SECONDS);
+    var toReactivate = executables.get(uuid);
+    if (toReactivate == null) {
+      LOG.error("Inbound connector executable not found for reactivation, received ID: {}", uuid);
+      return;
+    }
+    this.reactivationExecutor.schedule(
+        () -> tryRestart(uuid, toReactivate, delay, retryAttempts),
+        delay.getSeconds(),
+        TimeUnit.SECONDS);
+  }
+
+  private void tryRestart(
+      UUID uuid, RegisteredExecutable toReactivate, Duration delay, Integer retryAttempts) {
+    if (retryAttempts < 0) return;
```

Comment:
What is the specification here? We stop restart if the the attempts are for example -1? Its not clear when this would happen

Reply:
Here depends of what we want to do:

- In this case, `retryAttempts` means that if the user has set 3 retries, It means it will initially try then retry 3 times
- In the other case, if the user has set 3 retries, it will try initially and then try again 2 times

Reply:
I am not sure but `retryAttempts` sounds like the number of attempts that already happened. I would prefer a sealed interface with first of two options: Just a restart and one with a delay.
Lets tackle the number of retries later.